### PR TITLE
frontend: Add graph model, grouping and filtering

### PR DESCRIPTION
**Pull request by sniok**
Create the base graph model and basic operations like grouping and filtering for the resource map

Part of https://github.com/headlamp-k8s/headlamp/issues/198 and https://github.com/headlamp-k8s/headlamp/pull/2225

### CODE REVIEW

File: frontend/src/components/resourceMap/graph/graphFiltering.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ * - `custom`: Filters nodes using a custom filter function provided in the filter
+ */
+export function filterGraph(nodes: GraphNode[], edges: GraphEdge[], filters: GraphFilter[]) {
+  const filteredNodes: GraphNode[] = [];
+  const filteredEdges: GraphEdge[] = [];
+
+  const visitedNodes = new Set();
+  const visitedEdges = new Set();
+
+  function pushRelatedNodes(node: GraphNode) {
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphFiltering.ts:43]: Documentation please?

File: frontend/src/components/resourceMap/graph/graphFiltering.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      id: string;
+    }
+  | {
+      type: 'custom';
+      label: string;
+      filterFn: (node: GraphNode) => boolean;
+    };
+
+/**
+ * Filters the graph nodes and edges based on the provided filters
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphFiltering.ts:27]: Can you please add in the docs if it does AND or OR? From the code it seems it matches any of the filters, not all the filters.

Something like?
> * The filters are applied using an OR logic, meaning that a node will be included if it matches any of the filters.

File: frontend/src/components/resourceMap/graph/graphFiltering.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * The filters can be of the following types:
+ * - `name`: Filters nodes by the name
+ * - `related`: Keeps only the node with the id and nodes connected to it with edges
+ * - `hasErrors`: Filters nodes that have errors based on their resource status. See {@link getStatus}
+ * - `namespace`: Filters nodes by their namespace
+ * - `custom`: Filters nodes using a custom filter function provided in the filter
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphFiltering.ts:34]: The params could be documented too.

File: frontend/src/components/resourceMap/graph/graphFiltering.test.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    const { nodes: filteredNodes } = filterGraph(nodes, edges, filters);
+    expect(filteredNodes.map(it => it.id)).toEqual(['1', '2']);
+  });
+
+  it('filters nodes by custom filter function', () => {
+    const filters: GraphFilter[] = [
+      { type: 'custom', label: 'custom', filterFn: node => node.id === '1' },
+    ];
+    const { nodes: filteredNodes } = filterGraph(nodes, edges, filters);
+    expect(filteredNodes.map(it => it.id)).toEqual(['1', '2']);
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphFiltering.test.ts:66]: Can you please explain this test?

Maybe it's a bug. I would assume a filter for `id === '1'` would return only the id of '1'.

**Review by illume — CHANGES_REQUESTED** [review]
I reviewed filterGraph and left some notes.

File: frontend/src/components/resourceMap/graph/graphGrouping.test.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  it('groups nodes by namespace', () => {
+    const groupedGraph = groupGraph(nodes, edges, { groupBy: 'namespace' });
+    const namespaces = groupedGraph.data.nodes.map(node => node.id);
+    expect(namespaces).toEqual(['Namespace-ns1', 'Namespace-ns2']);
+  });
+
+  it('groups nodes by node', () => {
+    const groupedGraph = groupGraph(nodes, edges, { groupBy: 'node' });
+    const nodeNames = groupedGraph.data.nodes.map(node => node.id);
+    expect(nodeNames).toEqual(['Node-node1', '1', '3']);
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphGrouping.test.ts:70]: Can you please explain in this test why there are three nodes in the result, but four nodes in the test data? I would expect four nodes returned.

File: frontend/src/components/resourceMap/graph/graphGrouping.tsx [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+        const targetNode = nodes.find(n => n.id === edge.target);
+        if (targetNode) dfs(targetNode, componentNodes);
+      } else if (edge.target === node.id && !visited[edge.source]) {
+        const sourceNode = nodes.find(n => n.id === edge.source);
+        if (sourceNode) dfs(sourceNode, componentNodes);
+      }
+    });
+  };
+
+  nodes.forEach(node => {
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphGrouping.tsx:50]: A comment of what this function is doing would be good.

File: frontend/src/components/resourceMap/graph/graphGrouping.tsx [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+    const elseEdges = edges.filter(it => it.type !== 'kubeRelation');
+    root.data.edges.push(...elseEdges);
+    const groups = getConnectedComponents(nodes, relationEdges);
+    components = groups;
+  } else {
+    root.data.nodes = nodes;
+    root.data.edges = edges;
+  }
+
+  if (groupBy === 'namespace') {
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphGrouping.tsx:176]: A short comment explaining each of these blocks of logic would be appreciated.

File: frontend/src/components/resourceMap/graph/graphGrouping.tsx [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+/**
+ * Given a graph with groups, this function will 'collapse' all groups without
+ * the selected node. 'Collapsing' means that group won't show all children but
+ * only a preview
+ */
+export function collapseGraph(
+  graph: GroupNode | KubeGroupNode,
+  { selectedNodeId, expandAll }: { selectedNodeId?: string; expandAll: boolean }
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphGrouping.tsx:312]: Can you please explain this params?

File: frontend/src/components/resourceMap/graph/graphGrouping.tsx [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+    }
+  }
+
+  return undefined;
+}
+
+/**
+ * Given a graph with groups, this function will 'collapse' all groups without
+ * the selected node. 'Collapsing' means that group won't show all children but
+ * only a preview
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphGrouping.tsx:308]: Please add a `returns` explanation?

File: frontend/src/components/resourceMap/graph/graphGrouping.tsx [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+  { selectedNodeId, expandAll }: { selectedNodeId?: string; expandAll: boolean }
+) {
+  let root = { ...graph };
+  let selectedGroup: GraphNode | undefined;
+
+  if (selectedNodeId) {
+    selectedGroup = findGroupContaining(graph, selectedNodeId);
+  }
+
+  const collapseGroup = (group: GraphNode): GraphNode => {
```
Comment (illume) [inline frontend/src/components/resourceMap/graph/graphGrouping.tsx:321]: This function could use some documentation.

**Review by illume — CHANGES_REQUESTED** [review]
Reviewed graphGrouping and left some notes.

**Comment by illume** [comment]
Please remember to add a PR description, including a link to the issue.

File: frontend/src/components/resourceMap/graph/graphFiltering.test.ts (hunk 4 above)
Comment (sniok) [reply frontend/src/components/resourceMap/graph/graphFiltering.test.ts:74]: filterGraph also includes related nodes, so this is expected. added a comment to both filterGraph and inside tests

I now realize the the "related" filter should have been named "id" filter instead but at this point this change can be implemented in later PR

File: frontend/src/components/resourceMap/graph/graphGrouping.test.ts (hunk 5 above)
Comment (sniok) [reply frontend/src/components/resourceMap/graph/graphGrouping.test.ts]: added comments here

**Comment by sniok** [comment]
Implemented fixes for the comments, also updated the tests a bit to be more clearer

**Review by illume — APPROVED** [review]
🎉 thanks

**Pull request merged**